### PR TITLE
HAAR-887 Add endpoint to find users in multiple authsources

### PR DESCRIPTION
**Pull request by karenmillermoj**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/adapter/auth/AuthApiService.kt [hunk 1]
```diff
@@ -71,8 +75,31 @@ class AuthApiService(
...
+    status: Status?,
+    authSources: List<AuthSource>?,
+    page: Int?,
+    size: Int?,
+    sort: String?,
+  ) =
+    userWebClientUtils.getWithParams(
+      "/api/user/search",
+      object : ParameterizedTypeReference<PagedResponse<ExternalUserDetailsDto>>() {},
+      mapNonNull(
```
Comment (karenmillermoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/adapter/auth/AuthApiService.kt:90]: Added mapNonNull - not sure if we need it but if the endpoint isn't called with it, it's not passed on. Possibly discuss?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/external/UserRolesController.kt [hunk 2]
```diff
@@ -191,7 +191,7 @@ class UserRolesController(
   ) = userRolesService.addRolesByUserId(userId, roles)
 
   @PreAuthorize("hasAnyRole('ROLE_MAINTAIN_OAUTH_USERS', 'ROLE_AUTH_GROUP_MANAGER')")
-  @GetMapping("{userId}/assignable-roles")
+  @GetMapping("/{userId}/assignable-roles")
```
Comment (karenmillermoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/external/UserRolesController.kt:194]: added to be consistent

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/UserSearchController.kt [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  private val userSearchService: UserSearchService,
+) {
+
+  @GetMapping("/users/search")
+  @Operation(
+    summary = "Search for users ",
+    description = """
+      Search for users in the Auth DB only who match on partial first name, surname, username or email and return a pageable result set. 
+      Optionally choose the authentication sources from any combination of auth, delius, nomis and azuread sources.
+      It will default to AuthSource.auth if the authSources parameter is omitted.
```
Comment (AlTurner-MOJ) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/UserSearchController.kt:29]: One to discuss, but wondering whether we should carry AuthSource.auth forward into this API? i.e. would it be better for this API to expose it as AuthSource.external? and provide an internal mapping from that to AuthSource.auth when calling Auth?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/integration/wiremock/HmppsAuthMockServer.kt [hunk 4]
```diff
@@ -210,4 +210,171 @@ class HmppsAuthMockServer : WireMockServer(WIREMOCK_PORT) {
...
+                  "numberOfElements": 0,
+                  "empty": true
+              }
+              """.trimIndent(),
+            ),
+        ),
+    )
+  }
+
+  fun stubUserSearchAllFiltersWithResults(name: String) {
```
Comment (AlTurner-MOJ) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/integration/wiremock/HmppsAuthMockServer.kt:257]: Should this all filters stub not include status and auth source too?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/UserSearchController.kt (hunk 3 above)
Comment (karenmillermoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/resource/UserSearchController.kt:29]: It won't let me comment on earlier change - very odd .
Yes - the mapNotNull should work - I have created a method mapNotNull....we can see it works as the api call that gets sent out to auth does not contain the additional parameters (in the controllerIntTest) - hope that's ok

File: src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/integration/wiremock/HmppsAuthMockServer.kt (hunk 4 above)
Comment (karenmillermoj) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/manageusersapi/integration/wiremock/HmppsAuthMockServer.kt:257]: yes - will update

**Review by AlTurner-MOJ — APPROVED** [review]
May need a separate ticket to look at AuthSource.auth. Other than that looks good to me

**Pull request merged**